pkg/utils/kube/resource_ops.go

```
@@ -125,6 +139,47 @@ func (k *kubectlResourceOperations) runResourceCommand(ctx context.Context, obj
 	return strings.Join(out, ". "), nil
 }
 
+func (k *kubectlServerSideDiffDryRunApplier) runResourceCommand(obj *unstructured.Unstructured, executor commandExecutor) (string, error) {
+	manifestBytes, err := json.Marshal(obj)
+	if err != nil {
+		return "", err
+	}
+	manifestFile, err := os.CreateTemp(io.TempDir, "")
+	if err != nil {
+		return "", fmt.Errorf("Failed to generate temp file for manifest: %v", err)
+	}
+	defer io.DeleteFile(manifestFile.Name())
+	if _, err = manifestFile.Write(manifestBytes); err != nil {
+		return "", fmt.Errorf("Failed to write manifest: %v", err)
+	}
+	if err = manifestFile.Close(); err != nil {
+		return "", fmt.Errorf("Failed to close manifest: %v", err)
+	}
+
+	err = maybeLogManifest(manifestBytes, k.log)
+	if err != nil {
+		return "", err
+	}
+
+	// Run kubectl apply
+	ioStreams := genericclioptions.IOStreams{
+		In:     &bytes.Buffer{},
+		Out:    &bytes.Buffer{},
+		ErrOut: &bytes.Buffer{},
+	}
+	err = executor(ioStreams, manifestFile.Name())
+	if err != nil {
+		return "", errors.New(cleanKubectlOutput(err.Error()))
+	}
+	if buf := strings.TrimSpace(ioStreams.ErrOut.(*bytes.Buffer).String()); len(buf) > 0 {
+		k.log.Info("Server-side dry run apply had non-empty stderr: %s", buf)
```

I think that if we get something in stderr, we should log and return the ErrOut as an error

---

Maybe as a warning, since the command itself didn't fail.

---

Ah, logr doesn't support warning

---

I'd rather keep it as info, errors can potentially result in spam. Looks like level 0 (info) is used for both info and warn, e.g. https://pkg.go.dev/github.com/go-logr/logr#ToSlogHandler.

---

My point was less about the log level and more about returning an error. Currently, even if the kubectl apply returns an error in stderr, we just log and continue with the function execution. It seems to me that instead we should return the error and stop the execution.

---

One example of what can be there is
```
Warning: resource configmaps/<placeholder> is missing the kubectl.kubernetes.io/last-applied-configuration annotation which is required by  apply.  apply should only be used on resources created declaratively by either  create --save-config or  apply. The missing annotation will be patched automatically.'
```
This seems like something that would be fixed automatically, so we may just proceed.

---

Will update as discussed offline.